contracts/StakeManager.sol

```diff
@@ -91,33 +121,41 @@ contract StakeManager is Ownable {
      *
      * @dev Reverts when `_time` is not in range of [MIN_LOCKUP_PERIOD, MAX_LOCKUP_PERIOD]
      */
-    function stake(uint256 _amount, uint256 _time) external onlyVault noMigration {
+    function stake(uint256 _amount, uint256 _time) external onlyVault noMigration processEpoch {
         if (_time > 0 && (_time < MIN_LOCKUP_PERIOD || _time > MAX_LOCKUP_PERIOD)) {
             revert StakeManager__InvalidLockupPeriod();
         }
         Account storage account = accounts[msg.sender];
-        processAccount(account, currentEpoch);
+        if (account.lockUntil == 0) {
+            // account not initialized
+            account.lockUntil = block.timestamp;
+            account.epoch = currentEpoch + 1; //starts next epoch
+            account.rewardAddress = StakeVault(msg.sender).owner();
+        } else {
+            processAccount(account, currentEpoch);
+        }
+        mintIntialMP(account, _time, _amount);
+        //update storage
+        totalSupplyBalance += _amount;
         account.balance += _amount;
-        account.rewardAddress = StakeVault(msg.sender).owner();
-        mintIntialMultiplier(account, _time, _amount, 1);
-        stakeSupply += _amount;
+        account.lockUntil += _time;
     }
 
     /**
-     * Decreases balance of msg.sender;
-     * @param _amount Amount of balance to be decreased
+     * leaves the staking pool and withdraws all funds;
      */
-    function unstake(uint256 _amount) external onlyVault noMigration {
+    function unstake() external onlyVault noMigration processEpoch {
```

Comment:
So we no longer allow for unstaking less than current stake amount?

@mart1n-xyz Double checking: is this desired?

Reply:
I removed because this function would give too much losses to user. It could be reimplemented if needed.

Reply:
I think we want to allow partial unstaking (with proportional slashing of MPs).